CMakeLists.txt

```diff
@@ -61,17 +61,24 @@ set(PACKAGE ${CMAKE_PROJECT_NAME})
 string(TOUPPER "${CMAKE_PROJECT_NAME}" UPCASE_PROJECT_NAME)
 
 # Top level cmake file, set options
-if(NOT NTTEC_USE_SIMD)
-  set(NTTEC_USE_SIMD "OFF"
+if((NTTEC_USE_SIMD STREQUAL "SSE4") OR (NTTEC_USE_SIMD STREQUAL "AVX2"))
+  set(NTTEC_USE_SIMD "${NTTEC_USE_SIMD}"
     CACHE
     STRING
     "Build with SIMD vectorizations."
     FORCE
   )
-  set_property(CACHE NTTEC_USE_SIMD PROPERTY STRINGS
-    "OFF" "SSE4" "AVX2"
+else()
+  set(NTTEC_USE_SIMD "OFF"
+    CACHE
+    STRING
+    "Build with SIMD vectorizations."
+    FORCE
   )
 endif()
+set_property(CACHE NTTEC_USE_SIMD PROPERTY STRINGS
+  "OFF" "SSE4" "AVX2"
+)
 add_definitions(-DSSE4=1)
```

Comment:
Why not defining these in the code directly?

Or even better: why not have "NTTEC_USE_AVX" and "NTTEC_USE_SSE" and set them accordingly, that way you can directly use `#if defined(NTTEC_USE_AVX)` instead of testing a numerical value.

e.g

```
# Option for enabling/disabling SIMD flags is for both of debug and release
if (NTTEC_USE_SIMD STREQUAL "SSE4")
  SET(COMMON_CXX_FLAGS "${COMMON_CXX_FLAGS} -msse4.1")
  add_definitions(-DNTTEC_USE_SSE)
elseif (NTTEC_USE_SIMD STREQUAL "AVX2")
  SET(COMMON_CXX_FLAGS "${COMMON_CXX_FLAGS} -mavx -msse4.2")
  add_definitions(-DNTTEC_USE_AVX)
endif()
```

Then, you don't need these definitions

Reply:
I'd like to keep this way since we could add vectorisations for AVX512

Reply:
The problem with your approach is that there is nothing to enforce the synchronization between

```
add_definitions(-DSSE4=1)
add_definitions(-DAVX2=2)
```

And things like `-DNTTEC_USE_SIMD=2`.

If you changes the definitions to 3, maybe you'll keep using 2 here and that will be buggy: this approach is error prone.
You won't have this kind of issue with what I suggest.
I don't see the problem with supporting AVX512? In both case, you'll have to add another else branch.